src/test/server/models/userModel-test.ts

```diff
@@ -0,0 +1,156 @@
+  t.deepEqual(result, null);
+});
+
+test.serial('should return data for existing user', async (t) => {
+  const exampleData = {
+    username: TEST_USER_TOKEN,
+    githubToken: '1234',
+    scopes: [],
+  };
+
+  await firestore()
+      .collection(USERS_COLLECTION_NAME)
+      .doc(TEST_USERNAME)
+      .set(exampleData);
+
+  const result = await userModel.getLoginDetails(TEST_USERNAME);
+  t.deepEqual(result, exampleData);
+});
+
+test.serial('should return a new user token', async (t) => {
+  const githubToken = 'fake-github-token';
+  const scopes = ['test-1', 'test-2'];
+
+  t.context.sandbox.stub(githubFactory, 'github').callsFake(() => {
+    return {
+      query: ({context}: {context: {githubToken: string}}) => {
+        t.deepEqual(context.githubToken, githubToken);
+        return {
+          data: {
+            viewer: {
+              login: TEST_USERNAME,
+            }
+          }
+        };
+      }
+    };
+  });
+
+  t.context.sandbox.stub(crypto, 'randomBytes').callsFake(() => {
+    return {
+      toString: () => {
+        return TEST_USER_TOKEN;
+      },
+    };
+  });
+
+  const newToken = await userModel.generateNewUserToken(githubToken, scopes);
+  t.deepEqual(newToken, TEST_USER_TOKEN);
+
+  const userDoc = await firestore()
+                      .collection(USERS_COLLECTION_NAME)
+                      .doc(TEST_USERNAME)
+                      .get();
+  t.deepEqual(userDoc.data(), {
+    githubToken,
+    username: TEST_USERNAME,
+    scopes,
+  });
+
+  const tokenDoc = await firestore()
+                       .collection(TOKEN_COLLECTION_NAME)
+                       .doc(TEST_USER_TOKEN)
+                       .get();
+  const data = tokenDoc.data();
+  if (!data) {
+    t.fail();
+    return;
+  }
+
+  t.deepEqual(data.username, TEST_USERNAME);
+});
+
+test.serial('should return null for no token', async (t) => {
+  const result = await userModel.getLoginFromToken(undefined);
+  t.deepEqual(result, null);
+});
+
+test('should return null for token that doesnt exit', async (t) => {
+  const result = await userModel.getLoginFromToken(TEST_USER_TOKEN);
+  t.deepEqual(result, null);
+});
+
+test.serial('should return null for token that has no username', async (t) => {
+  await firestore()
+      .collection(TOKEN_COLLECTION_NAME)
+      .doc(TEST_USER_TOKEN)
+      .set({});
+
+  const result = await userModel.getLoginFromToken(TEST_USER_TOKEN);
+  t.deepEqual(result, null);
+});
+
+test.serial(
+    'should return null for token that has username but no user info',
+    async (t) => {
+      await firestore()
+          .collection(TOKEN_COLLECTION_NAME)
+          .doc(TEST_USER_TOKEN)
+          .set({
+            username: TEST_USERNAME,
+          });
+
+      const result = await userModel.getLoginFromToken(TEST_USER_TOKEN);
+      t.deepEqual(result, null);
+    });
+
+test.serial('should return logindetails for token', async (t) => {
```

Comment:
login details

Reply:
I'm not sure this is a great test. It's very tightly coupled with the implementation instead of testing the high level API. I would prefer if this final test `getLoginFromToken()` was moved to the end of `'should return a new user token'`.